### DISCUSSION TITLE
When I used the delay plug-in, I started the corresponding message confirmation and found that the message replied to the ACK message to the switch, but the switch made an error to the corresponding queue,Tips：replyText:NO_ROUTE

### DISCUSSION THREAD

**Discussion by a1145798605** [rabbitmq/rabbitmq-delayed-message-exchange#175] [General]
TraceId=20210902990583 - 2021-09-02 16:43:54.442|http-nio-8086-exec-6|INFO |com.code.springboot.zhoujl.common.interceptor.HttpBaseInterceptor|preHandle|32 -> 开始访问:http://localhost:8086/test/sendDelayMsg
TraceId=20210902990583 - 2021-09-02 16:43:54.443|http-nio-8086-exec-6|INFO |com.code.springboot.zhoujl.common.interceptor.HttpBaseInterceptor|preHandle|33 -> ip地址:172.35.0.18
msg= 我来发一个测试消息,10秒.delayTime10000
2021-09-02 16:43:54 Delay sent.
msg= 我来发一个测试消息，20秒.delayTime20000
2021-09-02 16:43:54 Delay sent.
msg= 我来发一个测试消息，30秒.delayTime30000
2021-09-02 16:43:54 Delay sent.
 - 2021-09-02 16:43:54.498|connectionFactory3|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|returnedMessage|105 -> Switch message not queued  replyText:NO_ROUTE,exchange:exchange.xdelay.delayed,routingKey:routingkey.xdelay.delay,消息:"我来发一个测试消息,10秒"
 - 2021-09-02 16:43:54.498|connectionFactory4|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|confirm|88 -> rabbitmq消息id:CorrelationData [id=acac907f612e4cf1a49c88d8ac8c234d]
 - 2021-09-02 16:43:54.498|connectionFactory4|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|confirm|90 -> rabbitmq The message is sent to the switch to confirm success
 - 2021-09-02 16:43:54.499|connectionFactory3|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|returnedMessage|107 -> id:acac907f612e4cf1a49c88d8ac8c234d
 - 2021-09-02 16:43:54.499|connectionFactory5|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|returnedMessage|105 -> Switch message not queued replyText:NO_ROUTE,exchange:exchange.xdelay.delayed,routingKey:routingkey.xdelay.delay,消息:"我来发一个测试消息，20秒"
 - 2021-09-02 16:43:54.499|connectionFactory6|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|confirm|88 -> rabbitmq消息id:CorrelationData [id=8fa9221af2e947fe8a15aa78acecc54c]
 - 2021-09-02 16:43:54.500|connectionFactory5|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|returnedMessage|107 -> id:8fa9221af2e947fe8a15aa78acecc54c
 - 2021-09-02 16:43:54.500|connectionFactory6|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|confirm|90 -> rabbitmq The message is sent to the switch to confirm success
TraceId=20210902990583 - 2021-09-02 16:43:54.553|http-nio-8086-exec-6|INFO |com.code.springboot.zhoujl.common.interceptor.HttpBaseInterceptor|afterCompletion|43 -> 结束加载请求:http://localhost:8086/test/sendDelayMsg,总用时:109ms
 - 2021-09-02 16:43:54.596|connectionFactory6|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|returnedMessage|105 -> Switch message not queued replyText:NO_ROUTE,exchange:exchange.xdelay.delayed,routingKey:routingkey.xdelay.delay,消息:"我来发一个测试消息，30秒"
 - 2021-09-02 16:43:54.596|connectionFactory6|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|returnedMessage|107 -> id:fa2b898a0f0c4257839f7491a9a64beb
 - 2021-09-02 16:43:54.596|connectionFactory5|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|confirm|88 -> rabbitmq消息id:CorrelationData [id=fa2b898a0f0c4257839f7491a9a64beb]
 - 2021-09-02 16:43:54.597|connectionFactory5|INFO |zhoujl.rabbitmq.config.RabbitMqConfig|confirm|90 -> rabbitmq The message is sent to the switch to confirm success
收到延时消息时间：2021-09-02 16:44:04 Delay sent.
收到延时消息:"我来发一个测试消息,10秒"
收到延时消息时间：2021-09-02 16:44:14 Delay sent.
收到延时消息:"我来发一个测试消息，20秒"
收到延时消息时间：2021-09-02 16:44:24 Delay sent.
收到延时消息:"我来发一个测试消息，30秒"

News can be consumed successfully

//rabbitMq config  rabbitMq version 3.9 plus version 3.9
@Slf4j
@EnableRabbit
@Configuration
public class RabbitMqConfig implements RabbitListenerConfigurer {

    @Value("${spring.rabbitmq.host}")
    private String host;

    @Value("${spring.rabbitmq.port}")
    private int port;

    @Value("${spring.rabbitmq.username}")
    private String username;

    @Value("${spring.rabbitmq.password}")
    private String password;

    @Value("${spring.rabbitmq.virtual-host}")
    private String virtualHost;

    @Bean
    public ConnectionFactory connectionFactory() {
        CachingConnectionFactory connectionFactory = new CachingConnectionFactory();
        connectionFactory.setHost(host);
        connectionFactory.setPort(port);
        connectionFactory.setUsername(username);
        connectionFactory.setPassword(password);
        connectionFactory.setVirtualHost(virtualHost);
  
        connectionFactory.setPublisherConfirmType(CachingConnectionFactory.ConfirmType.CORRELATED);
    
        connectionFactory.setPublisherReturns(true);
        return connectionFactory;
    }

    @Bean
    public RabbitTemplate rabbitTemplate(@Autowired ConnectionFactory connectionFactory) {
        RabbitTemplate template = new RabbitTemplate(connectionFactory);
        template.setMessageConverter(new Jackson2JsonMessageConverter());
        template.setConfirmCallback(confirmCallback());
        template.setMandatory(true);
        template.setReturnCallback(returnCallback());
        return template;
    }

    @Bean
    @ConditionalOnMissingBean(value = RabbitTemplate.ConfirmCallback.class)
    public RabbitTemplate.ConfirmCallback confirmCallback() {
        return new RabbitTemplate.ConfirmCallback() {
            @Override
            public void confirm(CorrelationData correlationData, boolean ack, String cause) {
                log.info("rabbitmq消息id:{}",correlationData);
                if (ack) {
                    log.info("rabbitmq消息发送到交换机确认成功");
                } else {
                    log.info("rabbitmq消息发送到交换机确认失败:{}",cause);
                }
            }
        };
    }

    @Bean
    @ConditionalOnMissingBean(value = RabbitTemplate.ReturnCallback.class)
    public RabbitTemplate.ReturnCallback returnCallback() {
        return new RabbitTemplate.ReturnCallback() {
            @Override
            public void returnedMessage(Message message, int replyCode, String replyText, String exchange, String routingKey) {
            
                log.info("交换机消息没有到队列 replyText:{},exchange:{},routingKey:{},消息:{}",replyText,exchange,routingKey,
                        new String(message.getBody()));
                log.info("id:{}",message.getMessageProperties().getCorrelationId());
            }
        };
    }

    @Override
    public void configureRabbitListeners(RabbitListenerEndpointRegistrar registrar) {
        registrar.setContainerFactory(myRabbitListenerContainerFactory());
    }

    @Bean
    public SimpleRabbitListenerContainerFactory myRabbitListenerContainerFactory() {
        SimpleRabbitListenerContainerFactory simpleRabbitListenerContainerFactory = new SimpleRabbitListenerContainerFactory();
        simpleRabbitListenerContainerFactory.setConnectionFactory(connectionFactory());
        simpleRabbitListenerContainerFactory.setConcurrentConsumers(1);
        //simpleRabbitListenerContainerFactory.setPrefetchCount(0);
        simpleRabbitListenerContainerFactory.setMaxConcurrentConsumers(1);
        simpleRabbitListenerContainerFactory.setAcknowledgeMode(AcknowledgeMode.MANUAL);
        return simpleRabbitListenerContainerFactory;
    }
}
//Binding settings
@Configuration
public class RabbitConfig {
    public static final String IMMEDIATE_QUEUE_XDELAY = "queue.xdelay.immediate";
    public static final String DELAYED_EXCHANGE_XDELAY = "exchange.xdelay.delayed";
    public static final String DELAY_ROUTING_KEY_XDELAY = "routingkey.xdelay.delay";
    @Bean
    public Queue immediateQueue() {
        
        return new Queue(IMMEDIATE_QUEUE_XDELAY, true);
    }
    @Bean
    public CustomExchange delayExchange() {
        Map<String, Object> args = new HashMap<>();
        args.put("x-delayed-type", "direct");
        return new CustomExchange(DELAYED_EXCHANGE_XDELAY, "x-delayed-message", true, false, args);
    }
    @Bean
    public Binding bindingNotify() {
        return BindingBuilder.bind(immediateQueue()).to(delayExchange()).with(DELAY_ROUTING_KEY_XDELAY).noargs();
    }
}

Thank you

**Comment by michaelklishin** [comment]
I will convert this issue to a GitHub discussion. Currently GitHub will [automatically close and lock the issue](https://github.com/github/feedback/discussions/3197) even though your question will be transferred and responded to elsewhere. This is to let you know that we do not intend to ignore this but this is how the current GitHub conversion mechanism makes it seem for the users :(

**Comment by michaelklishin** [comment]
This plugin does not support [mandatory publishing](https://github.com/rabbitmq/rabbitmq-delayed-message-exchange/issues/138) and returns. This is [explicitly mentioned in the Limitations section](https://github.com/rabbitmq/rabbitmq-delayed-message-exchange#limitations). Mandatory publishing and delayed publishing in combination make no sense: there is no way the plugin
can guarantee that there will be any bound queues to route to in the future.

**Reply by a1145798605** [reply]
Thank you for your answer

**Reply by bingtianyiyan** [reply]
I also encountered this problem，Mandatory =false. Prometheus gave an alarm and the message could not be routed. In fact, the message was sent to the delay queue normally. Can we consider improving this issue? Normal entry into the delay queue is considered successful, rather than routing failure

**Reply by michaelklishin** [reply]
@bingtianyiyan this exchange does not publish anything immediately, so it affects the unroutable messages metric.

You are welcome to contribute to this plugin but all contributions should go into https://github.com/rabbitmq/rabbitmq-delayed-message-exchange/pull/253 #229. The current version is no longer worked on.
